### PR TITLE
Use sts to construct stream arn

### PR DESCRIPTION
**Pull request by zengyu714**
*Description of changes:*
1. In order for users to benefit from account endpoints, StreamARN is included into both ListShards and PutRecords requests
1. Updated few dependencies to the most recent versions: aws-sdk-cpp. curl, and openssl, zlib
1. Added the STS dependency for building the StreamARN
   - Include STS endpoint and STS port in the configuration, and protobuf is updated accordingly
   - When the STS call fails, the KPL would exit. Therefore we can assure that KPL is running with StreamARNs
1. Fixed issue 
  - #445
  - #452

*Testing*
- Compiled in OSX, and tested with `amazon-kinesis-producer-sample`

```
[kpl-daemon-0003] INFO com.amazonaws.services.kinesis.producer.LogInputStreamReader - [2022-12-19 10:55:35.409865] [0x0000fe11][0x0000700009aea000] [info] [AWS Log: WARN](STSAssumeRoleWithWebIdentityCredentialsProvider)Token file must be specified to use STS AssumeRole web identity creds provider.
[kpl-daemon-0003] INFO com.amazonaws.services.kinesis.producer.LogInputStreamReader - [2022-12-19 10:55:36.023747] [0x0000fe11][0x0000700009aea000] [info] [pipeline.h:218] StreamARN "arn:aws:kinesis:us-east-1:448057645220:stream/kpltest" has been successfully configured, and will be used in requests including ListShards and PutRecords
[kpl-daemon-0003] INFO com.amazonaws.services.kinesis.producer.LogInputStreamReader - [2022-12-19 10:55:36.025112] [0x0000fe11][0x0000700009aea000] [info] [shard_map.cc:89] Updating shard map for stream "kpltest"
[pool-1-thread-1] INFO com.amazonaws.services.kinesis.producer.sample.SampleProducer - Put 2000 of 200000 so far (1.00 %), 0 have completed (0.00 %)
[pool-1-thread-1] INFO com.amazonaws.services.kinesis.producer.sample.SampleProducer - Oldest future as of now in millis is 994
[kpl-daemon-0003] INFO com.amazonaws.services.kinesis.producer.LogInputStreamReader - [2022-12-19 10:55:36.626632] [0x0000fe11][0x000070000a191000] [info] [shard_map.cc:152] Successfully updated shard map for stream "kpltest" (arn: "arn:aws:kinesis:us-east-1:448057645220:stream/kpltest"). Found 4 shards
[pool-1-thread-1] INFO com.amazonaws.services.kinesis.producer.sample.SampleProducer - Put 4000 of 200000 so far (2.00 %), 0 have completed (0.00 %)
```
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/kinesis/core/pipeline.h [hunk 1]
```diff
@@ -190,8 +200,34 @@ class Pipeline : boost::noncopyable {
...
+    Aws::STS::Model::GetCallerIdentityRequest request;
+    auto outcome = sts.GetCallerIdentity(request);
+    if (outcome.IsSuccess()) {
+      auto result = outcome.GetResult();
+      Aws::Utils::ARN sts_arn(result.GetArn());
+
+      // Construct and return the Kinesis stream ARN.
+      std::stringstream arn;
+      arn << "arn:" << sts_arn.GetPartition() << ":kinesis:" << region << ":" << result.GetAccount()
+          << ":stream/" << stream_name;
```
Comment (shanmsac) [inline aws/kinesis/core/pipeline.h:215]: Is this stream arn format same for all regions? I thought we had some issues with arn format in some regions?

Comment (zengyu714) [reply aws/kinesis/core/pipeline.h:215]: That's a good call out, we once had a problem with pod1, but I think the format is good for all public commercial regions

**Review by shanmsac — APPROVED** [review]
lgtm

File: aws/kinesis/core/pipeline.h (hunk 1 above)
Comment (yasemin-amzn) [reply aws/kinesis/core/pipeline.h:215]: yep, we should still test to make sure this works in all partitions.

File: aws/kinesis/core/pipeline.h [hunk 2]
```diff
@@ -190,8 +200,34 @@ class Pipeline : boost::noncopyable {
...
+
+      auto arn_str = arn.str();
+      LOG(info) << "StreamARN \"" << arn_str << "\" has been successfully configured, "
+                << "and will be used in requests including ListShards and PutRecords";
+      return arn_str;
+    }
+
+    LOG(warning) << "Failed to get StreamARN using STS GetCallerIdentity with exception: "
```
Comment (yasemin-amzn) [inline aws/kinesis/core/pipeline.h:223]: we should throw if this fails. 

that will avoid operating in dual-mode, and have all applications start using ARN from this version onwards.

**Review by shanmsac — APPROVED** [review]
lgtm

**Comment by Dhara18** [comment]
Hi I am getting error like this
message: [2024-12-04 11:40:09.353630] [0x000015b5][0x00007f3fd99fc640] [error] [pipeline.h:230] Failed to get StreamARN using STS GetCallerIdentity | Code:  | Message: curlCode: 7, Couldn't connect to server | Request was: Action=GetCallerIdentity&Version=2011-06-15	
is it because i am missing either STS endpoint or STS port in the configuration ?

**Pull request merged**